src/poll.rs

```diff
@@ -1,60 +1,134 @@
-use failure::Error;
-use futures::{Async, Poll};
-use socket::MioSocket;
+use std::task::Context;
 
+use futures::{ready, Poll};
 use mio::Ready;
-use tokio::reactor::PollEvented2;
-
+use tokio::reactor::PollEvented;
 use zmq;
 
-pub trait Poller {
-    fn send_message(&self, msg: &zmq::Message) -> Poll<(), Error>;
+use crate::socket::SocketWrapper;
+use crate::{Multipart, Result};
 
-    fn recv_message(&self, msg: &mut zmq::Message) -> Poll<(), Error>;
-}
+/// Wrapper on top of a ZeroMQ socket, implements functions for asynchronous reading and writing
+/// of multipart messages.
+pub(crate) struct EventedSocket(pub(crate) PollEvented<SocketWrapper>);
+
+impl EventedSocket {
+    /// Creates a new `EventedSocket` from a ZeroMQ socket.
+    pub(crate) fn from_zmq_socket(socket: zmq::Socket) -> Self {
+        Self(PollEvented::new(SocketWrapper::new(socket)))
+    }
+
+    /// Attempt to receive a Multipart message from a ZeroMQ socket.
+    ///
+    /// Either the whole multipart at once or nothing is received (according to
+    /// [http://zguide.zeromq.org/page:all#Multipart-Messages], when one part of a multipart message
+    /// has been received, all the others are already available as well).
+    ///
+    /// If nothing was received, the read flag is cleared.
+    pub(crate) fn multipart_recv(&mut self, cx: &mut Context) -> Poll<Option<Result<Multipart>>> {
+        let ready = Ready::readable();
+        ready!(self.0.poll_read_ready(cx, ready))?;
 
-impl Poller for PollEvented2<MioSocket> {
-    fn send_message(&self, msg: &zmq::Message) -> Poll<(), Error> {
-        match self.poll_write_ready()? {
-            Async::Ready(_) => {
-                //Send the message, and if it will block, then we set up a notifier
-                if let Err(e) = self.get_ref().io.send(&**msg, zmq::DONTWAIT) {
-                    if e == zmq::Error::EAGAIN {
-                        self.clear_write_ready()?;
-                        return Ok(Async::NotReady);
-                    } else {
-                        return Err(e.into());
+        let mut buffer = Multipart::new();
```

Comment:
Would this allocate a new VeqDeque if nothing was received?

Reply:
Good point, I also thought of that. Basically the tradeoff is:
1) Allocate upfront
- If there's no message, an allocation was wasted
+ for each message you just push

2) Allocate lazily
+ if there's no message, it's cheap
- for each message you have an additional branch

I don't actually know if VecDeque allocates a buffer right away, maybe it waits for the first item, I'll find out and benchmark both variants with different message sizes.

Reply:
`VecDeque` allocates with capacity 7 (https://doc.rust-lang.org/src/alloc/collections/vec_deque.rs.html#26), that is probably too much for most multiparts, I'll try to benchmark different configurations.